client/src/main/java/io/confluent/kafka/schemaregistry/rules/RuleContext.java

```diff
@@ -177,9 +179,30 @@ public FieldContext currentField() {
 
   public FieldContext enterField(Object containingMessage,
       String fullName, String name, RuleContext.Type type, Set<String> tags) {
-    Set<String> allTags = new HashSet<>(tags);
-    allTags.addAll(getTags(fullName));
-    return new FieldContext(containingMessage, fullName, name, type, allTags);
+    Set<String> metadataTags = getTags(fullName);
+    if (!metadataTags.isEmpty()) {
+      tags = new HashSet<>(tags);
+      tags.addAll(metadataTags);
+    }
+    Set<String> ruleTags = rule().getTags();
+    if (!type.isPrimitive() || ruleTags.isEmpty() || !disjoint(tags, ruleTags)) {
```

Comment:
Add optimization to check if null can be returned null

Reply:
don't know the logic well enough here but should this be `!ruleTags.isEmpty()`?

Reply:
Good question, if the rule tags are empty, then we apply the rule to all fields